File: src/mironov/exercise/cuda.cpp
#include <random>
#include <iomanip>
#include <array>
#include <exception>

#include <tvm/tvm.h>
#include <tvm/operation.h>
#include <tvm/tensor.h>
#include <tvm/build_module.h>
#include <topi/broadcast.h>

using namespace std;

int main(int argc, char **argv)
{
  /* Shape variable */
  auto n = tvm::var("n");
  tvm::Array<tvm::Expr> shape = {n};
  tvm::Tensor A = tvm::placeholder(shape, tvm::Float(32), "A");
  tvm::Tensor B = tvm::placeholder(shape, tvm::Float(32), "B");

  /* Build a graph for computing A + B */
  tvm::Tensor C = tvm::compute(shape, tvm::FCompute([=](auto i){ return A(i) + B(i); } )) ;

  /* Prepare a function `vecadd` with no optimizations */
  tvm::Schedule s = tvm::create_schedule({C->op});
  tvm::BuildConfig config = tvm::build_config();
  std::unordered_map<tvm::Tensor, tvm::Buffer> binds;
  auto lowered = tvm::lower(s, {A,B,C}, "vecadd", binds, config);

  /* Output IR dump to stderr */
  cerr << lowered[0]->body << endl;

  tvm::IterVar block_idx = tvm::thread_axis(tvm::Range(), "blockIdx.x");
  tvm::IterVar thread_idx = tvm::thread_axis(tvm::Range(), "threadIdx.x");

  tvm::IterVar i,j;
  s[C].split(C->op->root_iter_vars()[0],64,&i,&j);
  s[C].bind(i, block_idx);
  s[C].bind(j, thread_idx);

  /* Output IR dump to stderr */
  tvm::Target target = tvm::target::cuda();
  tvm::Target target_host = tvm::target::llvm();
  tvm::runtime::Module mod = tvm::build(lowered, target, target_host, config);
  mod->SaveToFile(std::string(argv[0]) + ".cuda", "cuda");
  return 0;
}

File: src/mironov/exercise/cuda.py
import tvm
import time
import numpy as np

device = "cuda"
suffix = "ptx"

n = tvm.var ("n")
A = tvm.placeholder ((n), name='A', dtype="float32")
B = tvm.placeholder ((n), name='B', dtype="float32")
C = tvm.compute (A.shape, lambda *i: A(*i) + B(*i), name='C')
s = tvm.create_schedule (C.op)

bx, tx = s[C].split (C.op.axis[0], factor=64)
s[C].bind(bx, tvm.thread_axis("blockIdx.x"))
s[C].bind(tx, tvm.thread_axis("threadIdx.x"))

module = tvm.build(s, [A, B, C], device, target_host="llvm")

temp = tvm.contrib.util.tempdir()
module.save (temp.relpath("myadd.o"))

# Save device code
module.imported_modules[0].save(temp.relpath("myadd.%s" %suffix))
# Create shared library
tvm.contrib.cc.create_shared(temp.relpath("myadd.so"), [temp.relpath("myadd.o")])

myadd = tvm.module.load (temp.relpath("myadd.so"))
# Import "deviced" code
myadd_device = tvm.module.load(temp.relpath("myadd.%s" %suffix))
# Import module
myadd.import_module(myadd_device)

ctx = tvm.context (device, 0)
n = 1024
a = tvm.nd.array(np.random.uniform(size=(n)).astype(A.dtype), ctx)
b = tvm.nd.array(np.random.uniform(size=(n)).astype(A.dtype), ctx)
c = tvm.nd.array(np.random.uniform(size=(n)).astype(A.dtype), ctx)
t0 = time.time()
myadd(a, b, c)
t1 = time.time()
print (device)
print ("GPU time: %s" %(t1 - t0))

